### ISSUE TITLE
Insert DICOM Tag

### ISSUE THREAD

**Issue by Subashree01** [dcmjs-org/dcmjs#89]
I need to Insert the new Tag Datas inside the DICOM files using Node js- dcmjs npm package. Currently I could able to insert the Tag's VR type, Tag ID, Tag Value and Tag Length, using the upsert tag.  Now i need to add the Tag Description also for the Tag Value. How can I do it?

 Currently I have used Below syntax in node API 

router.post("/updateDicomTagDatas", function(req,res,next)
{
    var WorkFlow= req.body.workFlow;
    var dcmEmail=req.body.dcmEmail;
    for(let i=0;i<allFilePaths.length;i++)
    {
    let arrayBuffer = fs.readFileSync(allFilePaths[i]).buffer;    
    let DicomDict = dcmjs.data.DicomMessage.readFile(arrayBuffer);
     DicomDict.upsertTag("00090114", "ST", dcmEmail);
     DicomDict.upsertTag("00090115", "ST", WorkFlow);
     let new_file_WriterBuffer = DicomDict.write();
     fs.writeFileSync(allFilePaths[i], Buffer.from(new_file_WriterBuffer));
    }
});

And if I do upsertTag, I get the tag value and other properties inserted but in Tag Description Property I get unkown Tag & Data

**Comment by pieper** [comment]
I'm afraid it's more complicated than that...

If you are looking at Tag Description on a system that doesn't know about your private tag (it isn't in the receiver's dicom data dictrionary) then it has no way of knowing the description.  Private tags are ones where the group is an odd number and for yours the group is 0009, so by default nobody but you knows what the values mean.  The most it can do is display the text.

Plus, adding private tags is more complex than just adding them to the dataset.  You can google for the details, but [here's a place to start](https://www.medicalconnections.co.uk/kb/DICOM-Private-Elements).

Once you understand how it should be laid out, you can create the required elements with dcmjs.  You have a look at the discussion here about adding standard elements to the dataset: https://github.com/dcmjs-org/dcmjs/issues/64

**Comment by Subashree01** [comment]
@pieper 
I have another Question, whenever I tried to write the tag datas inside the DICOM file using dcmjs npm. I can see that the rest of the tag datas are getting modified. I don't why this happens.

**Comment by pieper** [comment]
Sounds like this is related to your questions in #90.  Your testing is very helpful, so we should try to get this sorted out.  Can you try creating a simple example script + dataset to use for debugging?

**Comment by rvantklooster** [comment]
I am running into the same kind of problem. I want to use dcmjs to read DICOM files, make some change to the headers and then write the new DICOM file to disk or blob.

I observed that Decimal Strings are not written properly to the output DICOM. I did some debugging and I noticed two things (example using tag (0018,0050), SliceThickness, DS, with value of 4.8):
- BufferStream::writeString is used to write the value. It correctly extracts the decimal 4.8 from the input. Then uses toUTF8Array(str) to convert the value 4.8, but 4.8 is treated as string so toUTF8Array(value) returns an empty array. Resulting in not writing the value to file.
- Another observation, the value of a decimal String is read in correctly, put but stored as array. The value stored is _[4.8]_, I would assume _4.8_ . Maybe it is a design choice, but for me it was surprising.
